tools/deploy_firmware.py

```diff
@@ -0,0 +1,288 @@
+        raise NotImplementedError
+
+
+class NetgearRax40(PrplwrtDevice):
+    """A Netgear rax40.
+    At the moment, the rax40 can only be updated through uboot using a tftp server.
+    """
+
+    tftp_dir = "/srv/tftp"
+    """The root directory of the tftp server. The image will be copied there"""
+
+    uboot_prompt = "GRX500 #"
+    """The u-boot prompt on the target."""
+
+    def upgrade_uboot(self):
+        """Upgrade the device through u-boot.
+
+        It requires a running tftp server listenning on the IP the
+        target device will expect (see `ipaddr` in the uboot environment).
+        """
+        serial_path = "/dev/{}".format(self.name)
+        if not os.path.exists(serial_path):
+            raise ValueError("The serial device {} does not exist!\n".format(serial_path)
+                             + "Please make sure you have an appropriate udev rule for it.")
+        print("Copying image '{}' to '{}'".format(self.image, self.tftp_dir))
+        shutil.copy(os.path.join(self.artifacts_dir, self.image), self.tftp_dir)
+        print("Image copied to {}.".format(self.tftp_dir))
+        with serial.Serial(serial_path, self.baudrate) as ser:
+            print("Connecting to serial")
+            shell = pexpect.fdpexpect.fdspawn(ser, logfile=sys.stdout.buffer)
+            if not shell.isalive():
+                raise ValueError("Unable to connect to the serial device!")
+            print("Connected")
+            # The console might not be active yet:
+            shell.sendline("")
+            # make the shell prompt appear:
+            self.set_prompt(shell)
+            shell.expect(self.shell_prompt)
+            # reboot:
+            shell.sendline("reboot")
+            shell.expect(["Hit any key to stop autoboot:",
+                          pexpect.EOF, pexpect.TIMEOUT], timeout=120)
+            # stop autoboot:
+            shell.sendline("")
+            shell.expect(self.uboot_prompt)
+            # do the actual upgrade:
+            shell.sendline("run update_fullimage")
+            shell.expect(self.uboot_prompt, timeout=600)
+            # reboot:
+            shell.sendline("reset")
+            shell.expect("Please press Enter to activate this console", timeout=180)
+            # activate the console:
+            shell.sendline("")
+            self.set_prompt(shell)
+            shell.expect(self.shell_prompt)
+            shell.sendline("exit")
+        if not self.reach(attempts=10):
+            raise ValueError("The device was not reachable after the upgrade!")
+
+
+class Generic(PrplwrtDevice):
+    """A generic PrplwrtDevice.
+
+    Since upgrading through uboot is generally target-specific, it
+    only offers the sysupgrade option.
+    """
+
+    def sysupgrade(self):
+        serial_path = "/dev/{}".format(self.name)
+        if not os.path.exists(serial_path):
+            raise ValueError("The serial device {} does not exist!\n".format(serial_path)
```

Comment:
I agree. 
I only added the serial connection for the "generic" targets later on and missed this.
I'll also keep this for later however, because I really need this PR to test the new prplWrt release candidate.